File: algorithms/0239.sliding_window_maximum/sliding_window_maximum.cpp
class Solution {
public:
    class MonoQueue {
    public:
        void pop(const int value) {
            if (!data_.empty() && data_.front() == value) {
                data_.pop_front();
            }
        }
        void push(const int value) {
            while (!data_.empty() && data_.back() < value) {
                data_.pop_back();
            }
            data_.push_back(value);
        }
        int front() const {
            return data_.front();
        }
    private:
        deque<int> data_;
    };
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        MonoQueue mono_queue;
        for (int i = 0; i < k; ++i) {
            mono_queue.push(nums[i]);
        }
        vector<int> max_values;
        max_values.push_back(mono_queue.front());
        for (int i = k; i < nums.size(); ++i) {
            mono_queue.pop(nums[i - k]);           
            mono_queue.push(nums[i]);
            max_values.push_back(mono_queue.front());
        }
        return max_values;
    }
};

File: algorithms/0239.sliding_window_maximum/sliding_window_maximum.rs
use std::collections::VecDeque;
impl Solution {

pub fn max_sliding_window(nums: Vec<i32>, window_size: i32) -> Vec<i32> {
    let window_size = window_size as usize;
    let mut maximums = vec![];
        let mut window = VecDeque::new();
    for (i, &n) in nums.iter().enumerate() {
        if !window.is_empty() && *window.front().unwrap() + window_size == i {
            window.pop_front();
        }
        while !window.is_empty() && nums[*window.back().unwrap()] < n {
            window.pop_back();
        }
        window.push_back(i);
        if i  >= window_size - 1 {
            maximums.push(nums[*window.front().unwrap()]);
        }
    }
    maximums
}
}
